paimon-core/src/main/java/org/apache/paimon/table/system/TagsTable.java

```diff
@@ -239,31 +240,24 @@ public RecordReader<InternalRow> createReader(Split split) {
                     CompoundPredicate compoundPredicate = (CompoundPredicate) predicate;
                     // optimize for IN filter
                     if ((compoundPredicate.function()) instanceof Or) {
-                        List<Predicate> children = compoundPredicate.children();
-                        for (Predicate leaf : children) {
-                            if (leaf instanceof LeafPredicate
-                                    && (((LeafPredicate) leaf).function() instanceof Equal
-                                            && ((LeafPredicate) leaf).literals().get(0)
-                                                    instanceof BinaryString)
-                                    && predicate
-                                                    .visit(LeafPredicateExtractor.INSTANCE)
-                                                    .get(TAG_NAME)
-                                            != null) {
-                                String equalValue =
-                                        ((LeafPredicate) leaf).literals().get(0).toString();
-                                if (tagManager.tagExists(equalValue)) {
-                                    predicateMap.put(equalValue, tagManager.tag(equalValue));
-                                }
-                            } else {
-                                predicateMap.clear();
-                                break;
-                            }
-                        }
+                        PredicateUtils.traverseCompoundPredicate(
+                                predicate,
+                                TAG_NAME,
+                                (Predicate p) -> {
+                                    String equalValue =
+                                            ((LeafPredicate) p).literals().get(0).toString();
+                                    if (predicateMap != null && tagManager.tagExists(equalValue)) {
+                                        predicateMap.put(equalValue, tagManager.tag(equalValue));
+                                    }
+                                },
+                                (Predicate p) -> {
+                                    predicateMap = null;
```

Comment:
Why not predicateMap.clear();

Reply:
had just set to null which more efficient.